File: helfomat-core/src/main/java/de/helfenkannjeder/helfomat/core/organization/event/OrganizationEditAddAttendanceTimeEvent.java
package de.helfenkannjeder.helfomat.core.organization.event;

import de.helfenkannjeder.helfomat.core.organization.AttendanceTime;
import de.helfenkannjeder.helfomat.core.organization.Organization;
import de.helfenkannjeder.helfomat.core.organization.OrganizationEventVisitor;
import de.helfenkannjeder.helfomat.core.organization.OrganizationId;

/**
 * @author Valentin Zickner
 */
@SuppressWarnings({"WeakerAccess", "CanBeFinal"})
public class OrganizationEditAddAttendanceTimeEvent extends OrganizationEditEvent {
    private int index;
    private AttendanceTime attendanceTime;

    protected OrganizationEditAddAttendanceTimeEvent() {
    }

    public OrganizationEditAddAttendanceTimeEvent(OrganizationId organizationId, int index, AttendanceTime attendanceTime) {
        super(organizationId);
        this.index = index;
        this.attendanceTime = attendanceTime;
    }

    public int getIndex() {
        return index;
    }

    public AttendanceTime getAttendanceTime() {
        return attendanceTime;
    }

    @Override
    public Organization.Builder applyOnOrganizationBuilder(Organization.Builder organization) {
        return organization.addAttendanceTime(index, attendanceTime);
    }

    @Override
    public <T> T visit(OrganizationEventVisitor<T> visitor) {
        return visitor.visit(this);
    }

}

File: helfomat-core/src/main/kotlin/de/helfenkannjeder/helfomat/core/organization/event/OrganizationEditAddAttendanceTimeEvent.kt
package de.helfenkannjeder.helfomat.core.organization.event

import de.helfenkannjeder.helfomat.core.organization.AttendanceTime
import de.helfenkannjeder.helfomat.core.organization.Organization
import de.helfenkannjeder.helfomat.core.organization.OrganizationEventVisitor
import de.helfenkannjeder.helfomat.core.organization.OrganizationId

/**
 * @author Valentin Zickner
 */
data class OrganizationEditAddAttendanceTimeEvent(
    override val organizationId: OrganizationId,
    val index: Int,
    val attendanceTime: AttendanceTime
) : OrganizationEditEvent(organizationId) {

    override fun applyOnOrganizationBuilder(organizationBuilder: Organization.Builder?, strictMode: Boolean): Organization.Builder? {
        if (organizationBuilder?.attendanceTimes != null) {
            if (organizationBuilder.attendanceTimes.size > index) {
                organizationBuilder.attendanceTimes.add(index, attendanceTime)
            } else {
                organizationBuilder.attendanceTimes.add(attendanceTime)
            }
        }
        return organizationBuilder
    }

    override fun <T> visit(visitor: OrganizationEventVisitor<T>): T {
        return visitor.visit(this)
    }

}